src/commands/goto.ts

```diff
@@ -87,9 +89,76 @@ function executeGoto(gotoType: number, editor: vscode.TextEditor, count: number,
     case 12: // go to last buffer modification position
       break
   }
+  return Promise.resolve()
+}
+
+function executeGotoLine(editor: vscode.TextEditor, count: number, extend: boolean, position: 'first' | 'end' | 'start' | 'default') {
+  const positions = {
+    'first': function (x: vscode.Selection) {
+      return editor.document.lineAt(x.active.line).firstNonWhitespaceCharacterIndex
+    },
+    'end': function (x: vscode.Selection) {
+      return editor.document.lineAt(x.active.line).range.end.character
+    },
+    'start': function (x: vscode.Selection) {
+      return 0
+    },
+    'default': function (x: vscode.Selection) {
+      return 0
+    },
+  }
+  
+  editor.selections = editor.selections.map(x => {
+      const npos = new vscode.Position(x.active.line, positions[position](x))
+      return new vscode.Selection(extend ? x.anchor : npos, npos)
+    }
+  )
+  return Promise.resolve()
+}
+
+function executeGotoDisplayLine(editor: vscode.TextEditor, count: number, extend: boolean, position: 'top' | 'center' | 'bottom' | 'default') {
+  const positions = {
+    'top': function () {
+      return editor.visibleRanges[0].start.line
+    },
+    'center': function () {
+      return (editor.visibleRanges[0].end.line + editor.visibleRanges[0].start.line) / 2
+    },
+    'bottom': function () {
+      return editor.visibleRanges[0].end.line
+    },
+    'default': function () {
+      return 0
+    },
+  }
+  
+  editor.selections = editor.selections.map(x => {
+      const npos = new vscode.Position(positions[position](), 0)
+      return new vscode.Selection(extend ? x.anchor : npos, npos)
+    }
+  )
+  return Promise.resolve()
+}
+
+function executeGotoFirstLine(editor: vscode.TextEditor, count: number, extend: boolean) {
+  const nanch = (extend
+                   ? editor.selections.map(x => x.anchor).reduce((prev, current) => (prev.isAfter(current) ? prev : current))
+                   : new vscode.Position(0, 0))
+  editor.selections = [new vscode.Selection(nanch, new vscode.Position(0, 0))]
+  return Promise.resolve()
+}
+
+function executeGotoLastLine(editor: vscode.TextEditor, count: number, extend: boolean, gotoLastChar: boolean = false) {
+  const lastline = (editor.document.lineCount > 0) ? (editor.document.lineCount - 1) : editor.document.lineCount
+  const npos = new vscode.Position(lastline, gotoLastChar ? editor.document.lineAt(lastline).range.end.character : 0)
+  const nanch = (extend
+                   ? editor.selections.map(x => x.anchor).reduce((prev, current) => (prev.isBefore(current) ? prev : current))
+                   : npos)
+  editor.selections = [new vscode.Selection(nanch, npos)]
+  return Promise.resolve()
```

Comment:
You're still returning `Promise`s here and at lines 92, 116 and 140.

Reply:
I have to return a promise because the ope file branch (case 11) in the switch statement.
Therefore at least for me it is more obvious when all functions return their own promise.

Would you mind if I remove the breaks and introduce a `return Promise.resolve()` again?

Reply:
Ah, that's right, I forgot about that. Never mind that comment! But the functions that don't use `Promise`s at all shouldn't return `Promise`s.